tests/torchtune/training/test_distributed.py

```diff
@@ -523,3 +523,223 @@ def _broadcast_full_state_dict(self, full_sd):
             result.append(None)
         torch.distributed.broadcast_object_list(result, src=0)
         return result[0]
+
+    @gpu_test(gpu_count=2)
+    @pytest.mark.skipif(
+        version.parse(torch.__version__).base_version < "2.5.0",
+        reason="torch >= 2.5 required",
+    )
+    def test_optimizer_in_backward(self):
```

Comment:
I'm wondering why we create a whole new test in here rather than adding a distributed test case in the existing [TestOptimInBackward test case](https://github.com/pytorch/torchtune/blob/5de500147e57d4d35e296faba68b296c03a1a2a0/tests/torchtune/training/test_memory.py#L91) (or if we want to inherit from `FSDPTest`, a new distributed version of that class in the same file). Because testing optimizer-in-backward as part of a class that is otherwise meant to test our fully_shard + state dict save and load logic feels a bit strange to me.

Reply:
The test here wants to proof that models running with optimzier-in-backward could have the same performance as running without.
The state_dict saving and loading want to test optimzier-in-backward's wrapper which is a bit different to the traditional optim.
There are not covered in `TestOptimInBackward`.

Reply:
If we want to do such an end-to-end test, maybe we can add a recipe test case instead? See e.g. test_full_finetune_distributed.py [here](https://github.com/pytorch/torchtune/blob/50d3ef13e709665b126904dcb7bae8e9679411f4/tests/recipes/test_full_finetune_distributed.py#L55-L66): with this test case you can run the full end-to-end recipe on a small test model and set `optimizer_in_bwd=True` directly from the config